core/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -133,15 +149,52 @@ public void execute(SearchContext searchContext) throws QueryPhaseExecutionExcep
             Collector collector;
             final Callable<TopDocs> topDocsCallable;
 
+            assert query == searcher.rewrite(query); // already rewritten
             if (searchContext.size() == 0) { // no matter what the value of from is
-                final TotalHitCountCollector totalHitCountCollector = new TotalHitCountCollector();
-                collector = totalHitCountCollector;
-                topDocsCallable = new Callable<TopDocs>() {
-                    @Override
-                    public TopDocs call() throws Exception {
-                        return new TopDocs(totalHitCountCollector.getTotalHits(), Lucene.EMPTY_SCORE_DOCS, 0);
+                while (true) {
+                    // remove wrappers that don't matter for counts
+                    if (query instanceof ConstantScoreQuery) {
+                        query = ((ConstantScoreQuery) query).getQuery();
+                    } else {
+                        break;
                     }
-                };
+                }
+                boolean canOptimizeCount =
+                        searchContext.terminateAfter() == SearchContext.DEFAULT_TERMINATE_AFTER
+                        && searchContext.parsedPostFilter() == null
+                        && searchContext.minimumScore() == null;
+                if (canOptimizeCount && query.getClass() == MatchAllDocsQuery.class) {
```

Comment:
It really really feels like this is something that should be bolted into the query rather than here. I know that isn't really an easy thing though.